Neos.ContentRepositoryRegistry/Classes/Service/EventMigrationService.php

```diff
@@ -367,6 +367,44 @@ public function migratePayloadToWorkspaceName(\Closure $outputFn): void
         $outputFn(sprintf('Migration applied to %s events.', $numberOfMigratedEvents));
     }
 
+    /**
+     * Rewrites all workspaceNames, that are not matching new constraints.
+     *
+     * Needed for feature "Stabilize WorkspaceName value object": https://github.com/neos/neos-development-collection/pull/5193
+     *
+     * Included in August 2024 - before final Neos 9.0 release
+     *
+     * @param \Closure $outputFn
+     * @return void
+     */
+    public function migratePayloadToValidWorkspaceNames(\Closure $outputFn): void
+    {
+        $backupEventTableName = DoctrineEventStoreFactory::databaseTableName($this->contentRepositoryId) . '_bkp_' . date('Y_m_d_H_i_s');
+        $outputFn('Backup: copying events table to %s', [$backupEventTableName]);
+        $this->copyEventTable($backupEventTableName);
+
+        $eventTableName = DoctrineEventStoreFactory::databaseTableName($this->contentRepositoryId);
+        $this->connection->beginTransaction();
+        $statement = <<<SQL
+                UPDATE {$eventTableName}
+                SET
+                  payload = JSON_SET(payload, '$.workspaceName', SUBSTR(MD5(JSON_UNQUOTE(JSON_EXTRACT(payload, '$.workspaceName'))), 1, 30))
```

Comment:
I read through this more in-depth now and I wonder why you used `SUBSTR(..., 1, ...)` – if the original value was uppercase it would still be invalid afterwards.
My original suggestion was:
```
CONCAT('w', MD5(JSON_UNQUOTE(JSON_EXTRACT(payload, '$.workspaceName')))))
```

(note the "w"-prefix)

Reply:
The md5 hash is 32 chars long, but the workspaceName can just be 30 chars long. So I had to shorten it. To reduce the chance of collision, I removed the prefix to be able to keep 30 chars of the hash instead of 29.

But yeah...